metrics/metrics_collector.go

```diff
@@ -0,0 +1,160 @@
+package metrics
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type metricCollector struct {
+	counters   []prometheus.Counter
+	gauges     []prometheus.Gauge
+	histograms []prometheus.Histogram
+	summarys   []prometheus.Summary
+	interval   int
+}
+
+var (
+	promRegistry = prometheus.NewRegistry() // local Registry so we don't get Go metrics, etc.
+)
+
+func newMetricCollector() metricCollector {
+	return metricCollector{}
+}
+
+func (mc *metricCollector) updateCounter() {
+	for {
+		time.Sleep(time.Second * time.Duration(mc.interval))
+		for idx := 0; idx < len(mc.counters); idx++ {
+			mc.counters[idx].Add(rand.Float64())
+		}
+	}
+}
+
+func (mc *metricCollector) updateGauge() {
+	for {
+		time.Sleep(time.Second * time.Duration(mc.interval))
+		for idx := 0; idx < len(mc.gauges); idx++ {
+			mc.gauges[idx].Add(rand.Float64())
+		}
+	}
+}
+
+func (mc *metricCollector) updateHistogram() {
+	for {
+		time.Sleep(time.Second * time.Duration(mc.interval))
+
+		for idx := 0; idx < len(mc.histograms); idx++ {
+			lowerBound := math.Mod(rand.Float64(), 1)
+			increment := math.Mod(rand.Float64(), 0.05)
+			for i := lowerBound; i < 1; i += increment {
+				mc.histograms[idx].Observe(i)
+
+			}
+		}
+	}
+}
+func (mc *metricCollector) updateSummary() {
+	for {
+		time.Sleep(time.Second * time.Duration(mc.interval))
+
+		for idx := 0; idx < len(mc.summarys); idx++ {
+			lowerBound := math.Mod(rand.Float64(), 1)
+			increment := math.Mod(rand.Float64(), 0.05)
+			for i := lowerBound; i < 1; i += increment {
+				mc.summarys[idx].Observe(i)
+
+			}
+		}
+	}
+}
+
+func (mc *metricCollector) updateMetrics(count int) {
```

Comment:
This seems duplicative.  Perhaps extract the inner loop contents of the above methods into separate methods and re-use them here.  Or use higher-order programming:

```go
func (mc *metricCollector) updateCounter() {
  // One single loop here
  for _, c := range mc.counters {
    c.Add(rand.Float64())
  }
}

func updateLoop(update func(), delay time.Duration) {
  go func() {
    for {
      time.Sleep(delay)
      update()
    }
  }()
}
```

This can then be used to kick off a goroutine running any given function in an infinite loop:

```go
updateLoop(mc.updateCounter) // note that mc.updateCounter is passed as a value and not invoked here
```

For bonus points you could have `updateLoop` take a `context.Context` as its first parameter and watch its done channel to enable cleaning up the spawned goroutines when the `Context` is cancelled.  Could also use a `time.Ticker` instead of `time.Sleep` in that case, since you'd already have a `select` statement:

```go
t := time.Ticker(delay)
for {
  select {
  case <-ctx.Done():
    return
  case <-t.C:
    update()
  }
}
```